### PR TITLE
Add OpenTelemetry instrumentation for Model.stream

### PR DESCRIPTION
**Pull request by koxudaxi**
### TL;DR

Added OpenTelemetry instrumentation for `Model.stream()` to track streaming LLM calls.

### What changed?

- Extended OpenTelemetry instrumentation to support streaming LLM calls via `Model.stream()`
- Added a new `activate` parameter to the span context manager to control span activation
- Implemented proper span lifecycle management for streaming responses
- Added support for extracting response IDs and recording exceptions during streaming
- Created comprehensive test suite for streaming instrumentation
- Made the span context manager more robust by handling both activated and non-activated spans

### How to test?

1. Enable OpenTelemetry instrumentation:
   ```python
   from mirascope import llm
   llm.instrument_opentelemetry()
   ```

2. Use streaming with a model:
   ```python
   model = llm.Model(provider="openai:responses", model_id="gpt-4o")
   response = model.stream(messages=[llm.messages.user("What is 4200 + 42?")])
   for chunk in response:
       print(chunk, end="")
   ```

3. Verify spans are exported to your OpenTelemetry backend with proper attributes

### Why make this change?

This change completes the OpenTelemetry instrumentation by adding support for streaming LLM calls. Previously, only non-streaming calls were instrumented, leaving a gap in observability for streaming use cases. With this change, users can now track and monitor both streaming and non-streaming LLM interactions in their observability platforms, providing a complete picture of LLM usage, performance, and errors.

---

> [!NOTE]
> Add OpenTelemetry GenAI instrumentation for Model.stream with proper span lifecycle, response/error handling, and comprehensive tests.
> 
> - **Instrumentation (llm.py)**:
>   - Add streaming wrapper `Model.stream` via `_instrumented_model_stream` with span activation control (`activate=False`), using `otel_trace.use_span`, and iterator wrapping to close/end spans on completion or errors.
>   - Extend `_start_model_span` with `activate` flag; support manual span start/end when not activated.
>   - Record untracked params as event (`gen_ai.request.params.untracked`) and attach dropped param metadata.
>   - Make `_attach_response` resilient (`FormattableT | None`, safe `raw` access); capture `gen_ai.response.id` when available.
>   - Wire up `_wrap_model_stream`/`_unwrap_model_stream`; update `instrument_llm`/`uninstrument_llm` to manage streaming instrumentation.
> - **Tests**:
>   - Add integration tests and VCR cassettes for streaming: basic span emission, tools, JSON format, untracked params event, None params, response ID, tracer=None, without instrumentation, API error, and iterator error handling.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a3c2bcd96286897e041bc207e5297a0d547d0db8. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by koxudaxi** [comment]
* **#1572** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1572?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1569** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1569?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1563** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1563?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1562** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1562?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1561** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1561?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1560** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1560?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1559** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1559?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1557** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1557?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1552** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1552?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1531** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1531?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1514** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1514?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1513** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1513?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1506** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1506?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1505** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1505?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1504** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1504?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1503** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1503?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1503?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1551** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1551?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1496** <a href="https://app.graphite.com/github/pr/Mirascope/mirascope/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `v2`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python/mirascope/ops/_internal/instrument_llm.py [hunk 1]
```diff
@@ -211,13 +218,165 @@ def _unwrap_model_call() -> None:
...
+                response_obj,
+                request_messages=request_messages,
+            )
+        span_cm.__exit__(exc_type, exc, tb)
+
+    # Wrap the provider iterator so we can close the span when streaming ends.
+    def _wrapped_iterator() -> Iterator[StreamResponseChunk]:
+        try:
+            yield from chunk_iterator
+        except Exception as exc:  # noqa: BLE001
```
Comment (willbakst) [inline python/mirascope/ops/_internal/instrument_llm.py]: Shouldn't we catch a specific exception?

File: python/mirascope/ops/_internal/instrument_llm.py [hunk 2]
```diff
@@ -211,13 +218,165 @@ def _unwrap_model_call() -> None:
...
+
+@wraps(_ORIGINAL_MODEL_STREAM)
+def _instrumented_model_stream(
+    self: Model,
+    *,
+    messages: Sequence[Message],
+    tools: Sequence[Tool] | Toolkit | None = None,
+    format: FormatParam = None,
+) -> StreamResponse | StreamResponse[FormattableT]:
+    """Returns a GenAI-instrumented result of `Model.stream`."""
```
Comment (willbakst) [inline python/mirascope/ops/_internal/instrument_llm.py]: I'm a little confused by all the complexity here. Can't we just enter the span context inside of the iterator wrapper? That would only run once the iteration starts, right?

File: python/mirascope/ops/_internal/span.py [hunk 3]
```diff
@@ -112,16 +117,31 @@ def span(
         params=params,
     )
 
-    with tracer.start_as_current_span(
+    if activate:
```
Comment (willbakst) [inline python/mirascope/ops/_internal/span.py]: Can you explain this new `activation`​ argument you added? It looks like it does an early return without ending the span?

File: python/mirascope/ops/_internal/instrument_llm.py (hunk 1 above)
Comment (koxudaxi) [reply python/mirascope/ops/_internal/instrument_llm.py]: This is cleanup code that needs exception info for span recording; the exception is re-raised after cleanup, similar to  context manager's `__exit__`.

File: python/mirascope/ops/_internal/instrument_llm.py (hunk 2 above)
Comment (koxudaxi) [reply python/mirascope/ops/_internal/instrument_llm.py]: OK, I refactor it. It now start the span before Model.stream to catch request errors but only activate it inside the iterator so the scope opens when iteration begins, simplifying the wrapper.

File: python/mirascope/ops/_internal/span.py (hunk 3 above)
Comment (koxudaxi) [reply python/mirascope/ops/_internal/span.py]: activate toggles between auto-activating spans (call needs it) and creating dormant spans for streaming so we can scope them during iteration, and even in the non-activated path `__exit__` closes the span so nothing leaks.

File: python/mirascope/ops/_internal/instrumentation/llm/llm.py [hunk 4]
```diff
@@ -513,8 +536,156 @@ def _unwrap_model_call() -> None:
...
+            request_messages=messages,
+        )
+    except Exception as exc:  # pragma: no cover
+        span_cm.__exit__(type(exc), exc, exc.__traceback__)
+        raise
+
+    return response
+
+
+def _attach_stream_span_handlers(
```
Comment (willbakst) [inline python/mirascope/ops/_internal/instrumentation/llm/llm.py:605]: nit: same comment around method definition ordering

**Comment by willbakst** [comment]
### Merge activity

* **Dec 3, 5:26 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/Mirascope/mirascope/1503).
* **Dec 3, 5:28 PM UTC**: [Graphite](https://app.graphite.com/github/pr/Mirascope/mirascope/1503) rebased this pull request as part of a merge.
* **Dec 3, 5:29 PM UTC**: @willbakst merged this pull request with [Graphite](https://app.graphite.com/github/pr/Mirascope/mirascope/1503).
